File: shared/java/src/com/pebblefields/shared/ServerControl.java
package com.pebblefields.shared;

import java.io.Serializable;
import java.util.List;

import org.pebblefields.pebbleobject.Pebble;
import com.pebblefields.shared.interfaces.HandlesDataEvents;


/**
 * @author ken
 */
public class ServerControl extends Pebble implements Serializable{
	static public HandlesDataEvents handlesDataEvents;
	
	public ServerControl(Pebble dm) {
		super(dm);
	}
		
	public int executeMethod(String sp, Pebble pebble) {
		int ret = 1;
		if (sp.equals("doInit")) {
			this.doInit(pebble);
		} else if (sp.equals("getArrayItemDocument")) {
			this.getArrayItemDocument(pebble);
		} else if (sp.equals("doQuery")) {
			this.doQuery(pebble);
		} else if (sp.equals("getFunctions")) {
			this.getFunctions(pebble);
		} else if (sp.equals("getFields")) {
			this.getFields(pebble);
		} else if (sp.equals("addItem")) {
			this.addItem(pebble);
		} else if (sp.equals("updateItem")) {
			this.updateItem(pebble);
		} else if (sp.equals("deleteItem")) {
			this.deleteItem(pebble);
		} else if (sp.equals("doQueries")) {
			this.doQueries(pebble);
		}
      return ret;
	}
	
	private void getArrayItemDocument(Pebble obj) {
		Pebble params = obj.getParams();
		String relPath = params.getRef("relPath");
		String collectionPath = params.getRef("collection");
		
		//check permissions
//		String collectionHoldingAppPath = collectionPath.substring(0, collectionPath.lastIndexOf("."));
//		String arrayField = collectionPath.substring(collectionPath.lastIndexOf(".") + 1);
//		PathAnalyzer collectionHoldingDoc = new PathAnalyzer(collectionHoldingAppPath);
//		Pebble collectionField = collectionHoldingDoc.getLastItemType().getField(arrayField);
//		
//		Pebble security = collectionField.get("security");
		
		
		Pebble doc = DocModel.getDoc(collectionPath + "." + relPath);
		
		obj.set("arrayItem", doc);
		
	} 
	/**
	 * Queries always query a path
	 * Return itemIndexes or appIndexes (appIndexes tell the client side to load as a app)
	 */
	private void doQuery(Pebble obj) {
		
		Pebble params = obj.getParams();
		Pebble query = params.get("query");

		if (query != null) {
			String collectionPath = query.getRef("path"); 
			PathAnalyzer collection = new PathAnalyzer(collectionPath);
			
			String collectionHoldingAppPath = collectionPath.substring(0, collectionPath.lastIndexOf("."));
			String arrayField = collectionPath.substring(collectionPath.lastIndexOf(".") + 1);
			PathAnalyzer collectionHoldingDoc = new PathAnalyzer(collectionHoldingAppPath);
			Pebble collectionField = collectionHoldingDoc.getLastItemType().getField(arrayField);

			Pebble type = collectionField.get("type");
			
			if (type.getBool("isArrayCollection")) {
				
				Pebble tmpDoc = new Pebble("<i><rs>" + getDocs(query) + "</rs></i>");
				obj.set(".", tmpDoc); //base class add the rs so we know we can replace it 
				
			} else {
				
				obj.set(".", collection.getItem());
				
			}
		}
	}
	private void getFunctions(Pebble obj) {
		
		Pebble params = obj.getParams();
		String controlId = params.getRef("controlId");

		ItemType control = new ItemType(controlId);
		Pebble functions = control.getFunctions();
		
		obj.set(".", functions);
		
	}
	private void getFields(Pebble obj) {
		
		Pebble params = obj.getParams();
		String typeId = params.getRef("typeId");

		ItemType control = new ItemType(typeId);
		Pebble fields = control.getFields();
		
		obj.set(".", fields);
		
	}

	private void doQueries(Pebble pebble) {
		Pebble params = pebble.getParams();
		List<Pebble> queries = params.getRecords("queries");

		Pebble type = new Pebble();
		type.setRef(".", TypeReferences.ARRAY);
		type.setRef("arrayFormId", TypeReferences.APP_INDEX_ITEM); //appIndexItem could be rename multipleQueryItem

		Pebble pn = new Pebble(pebble);
		//			pn.setNodeType(type);
		//			pn.setNodePath("MULTI_QUERIES");
		//			//pebble.setTrue("ps.nodeInfo.retrieved"); //use retrieve later

		String items = "";
		for (Pebble query : queries) {
//			String collectionPath = query.getRef("path");
//			CollectionInstance colInst = (CollectionInstance)ItemInstance.getServerInstance(collectionPath);

			//TODO: need a mapPath to appIndexItem.name, appIndexItem.ref is also required, or store together in different table...

			items += getDocs(query);
		}
		Pebble tmpDoc = new Pebble("<i><rs>" + items + "</rs></i>");
		pn.set(".", tmpDoc); //base class add the rs so we know we can replace it
	}	
	private String getDocs(Pebble queryItem) {
		
		String s = "";
		//List<String> dt = DocModel.getDatasource(queryItem.getRef("path")).doQuery(queryItem, null, null);
		//List<String> dt = DocModel.getDatasource().doQuery(queryItem, null, null);
		List<String> dt = DocModel.getDatasource().doQuery(queryItem);
		for (String dr : dt) {
			s += dr; //appIndexItem
		}
		return s;
	}
	 public void doInit(Pebble obj) {
		//set in request 
		Pebble user = obj.get("ps.user");
		obj.set("user", user);
		
	}
	//called as collection                      
    private void addItem(Pebble obj) {
    	Pebble params = obj.getParams();
        Pebble recItem = params.get("item");
        String collectionPath = params.getRef("collection");
        Pebble displayControl = params.get("displayControl.control");
        if (recItem != null ) {       
            String typePath = displayControl.getRef(".");
    		ItemType type = new ItemType(typePath);
    		//Pebble layout = type.get("layout", true);
    		//prunedItem = this.appInst.pruneItem(recItem, layout);
    				
    		String uniqueName = recItem.getValue("username");
            if (uniqueName != null) {
            	recItem.setTagName(uniqueName);
            }
            
            PathAnalyzer pathAnalyzer = new PathAnalyzer(collectionPath);
            ItemType parentForm = pathAnalyzer.getParentItemType();
            Pebble collectionField = parentForm.getField(collectionPath.substring(collectionPath.lastIndexOf(".") + 1));
            Pebble typeArrayMaps = collectionField.get("type.typeArrayMaps");
            String recType = collectionField.getRef("type.arrayFormId");
            ItemType form = new ItemType(recType);
            
            //Pebble doc = DocModel.getDatasource().create(collectionPath, recItem, typeArrayMaps, form);
            Pebble doc = DocModel.getDatasource().create(collectionPath, recItem);
            
            obj.set("appIndexItem", doc.get("ps.appIndexItem"));
            
            //notify
        }
    }
    
	/**
	 * Update any item instance data on the server 
	 * @param obj
	 */
	private void updateItem(Pebble obj) {
		Pebble params = obj.getParams();
        Pebble recItem = params.get("item");
        String itemPath = params.getRef("itemPath");
        String rPath = params.getRef("relPath");
        String collection = itemPath.substring(0, itemPath.lastIndexOf("."));
		String name = itemPath.substring(itemPath.lastIndexOf(".") + 1);
        Pebble displayControl = params.get("displayControl.control");
        if (recItem != null) {
        	PathAnalyzer pathAnalyzer = new PathAnalyzer(itemPath);
        	
        	String typePath = displayControl.getRef(".");
    		ItemType type = new ItemType(typePath);
    		//Pebble layout = type.get("layout", true);
    		//prunedItem = this.appInst.pruneItem(recItem, layout);
    		    		
//    		ArrayList<Pebble> items = prunedItem.getRecords(".");
//    		for (Pebble item : items) {
//    			String meta;
//    			if (path != null) {
//    				meta = path + "." + item.getMetaId();
//    			} else {
//    				meta = item.getMetaId();
//    			}
//    			obj.set(meta, item);
//    		}
    		
    		String relPath = pathAnalyzer.getRelPath();
    		Pebble obj1 = pathAnalyzer.getLastDoc();
    		obj1.set(relPath, recItem);
    		
    		PathAnalyzer pa = new PathAnalyzer(collection);
    		Pebble typeArrayMaps = pa.getLastFieldType().get("typeArrayMaps");
    		Pebble arrayForm = pa.getLastFieldType().get("arrayFormId");
    		
            //DocModel.getDatasource().update(collection, name, obj1, typeArrayMaps, new ItemType(arrayForm.getRef(".")));
    		DocModel.getDatasource().update(collection, name, obj1);
    		
    		String bPath = itemPath.replace("theInstance.theControlApp", "theControlItem");
    		String userId = obj.getRef("ps.user.id");
    		ServerControl.handlesDataEvents.notifyDataChange(recItem, bPath + "." + rPath, userId);         
    		
    		//TODO: pebble.set("appIndexItem", appIndexItem);
        }
    } 
	private void deleteItem(Pebble pebble) {
		Pebble params = pebble.getParams();
		String appPath = params.getRef("appPath");


		//	    	DocInstance app = (DocInstance)ItemInstance.getServerInstance(appPath, DocModel.UNAUTHORIZED_USER);
		//	    	app.deleteApp();
	}

	
	
	
}

File: shared/python/src/shared/servercontrol.py
from pebbleobject import pebble
import pathanalyzer
import docmodel
import itemtype

class ServerControl(pebble.Pebble):

	def __init__(self, dm) :
		pebble.Pebble.__init__(self, dm)


	def executeMethod (self, sp, obj) :
		ret = 1;
		if sp == "doInit" :
			self.doInit(obj);
		elif sp == "getArrayItemDocument" :
			self.getArrayItemDocument(obj);
		elif sp == "doQuery" :
			self.doQuery(obj);
		elif sp == "getFunctions" :
			self.getFunctions(obj);
		elif sp == "getFields" :
			self.getFields(obj);
		elif sp == "addItem" :
			self.addItem(obj);
		elif sp == "updateItem" :
			self.updateItem(obj);
		elif sp == "deleteItem" :
			self.deleteItem(obj);
		elif sp == "doQueries" :
			self.doQueries(obj);
		return ret;
	
	def getArrayItemDocument (self, obj) :
		params = obj.getParams();
		relPath = params.getRef("relPath");
		collectionPath = params.getRef("collection");
	
		#check permissions
	
	
		doc = docmodel.DocModel.getDoc(collectionPath + "." + relPath);
	
		obj.set("arrayItem", doc);
	
	def doQuery (self, obj) :
	
		params = obj.getParams();
		query = params.get("query");
	
		if query is not None :
			collectionPath = query.getRef("path"); 
			collection = pathanalyzer.PathAnalyzer(collectionPath);
	
			collectionHoldingAppPath = collectionPath[:collectionPath.rfind(".")];
			arrayField = collectionPath[collectionPath.rfind(".") + 1:];
			collectionHoldingDoc = pathanalyzer.PathAnalyzer(collectionHoldingAppPath);
			collectionField = collectionHoldingDoc.getLastItemType().getField(arrayField);
	
			type = collectionField.get("type");
	
			if type.getBool("isArrayCollection") :
	
				s = self.getDocs(query);
				tmpDoc = pebble.Pebble("<i><rs>" + s + "</rs></i>");
				obj.set(".", tmpDoc); #base class add the rs so we know we can replace it 
	
			else:
	
				obj.set(".", collection.getItem());
	
	def getFunctions (self, obj) :
	
		params = obj.getParams();
		controlId = params.getRef("controlId");
	
		control = itemtype.ItemType(controlId);
		functions = control.getFunctions();
	
		obj.set(".", functions);
	
	def getFields (self, obj) :
	
		params = obj.getParams();
		typeId = params.getRef("typeId");
	
		control = itemtype.ItemType(typeId);
		fields = control.getFields();
	
		obj.set(".", fields);
	
	def doQueries (self, obj) :
		params = obj.getParams();
		queries = params.getRecords("queries");
	
		type = pebble.Pebble();
	#	type.setRef(".", pebble.shared.TypeReferences.ARRAY);
	#	type.setRef("arrayFormId", pebble.shared.TypeReferences.APP_INDEX_ITEM); #appIndexItem could be rename multipleQueryItem
	
		pn = pebble.Pebble(obj);
		#			pn.setNodeType(type);
		#			pn.setNodePath("MULTI_QUERIES");
		#			#obj.setTrue("ps.nodeInfo.retrieved"); #use retrieve later
	
		items = "";
		for query in queries:
	#		String collectionPath = query.getRef("path");
	#		CollectionInstance colInst = (CollectionInstance)ItemInstance.getServerInstance(collectionPath);
	
			#TODO: need a mapPath to appIndexItem.name, appIndexItem.ref is also required, or store together in different table...
	
			items += self.getDocs(query);
		
		tmpDoc = pebble.Pebble("<i><rs>" + items + "</rs></i>");
		pn.set(".", tmpDoc); #base class add the rs so we know we can replace it
	
	def getDocs (self, queryItem) :
	
		s = "";
		dt = docmodel.DocModel.getDatasource(queryItem.getRef("path")).doQuery(queryItem, None, None);
		for dr in dt:
			s += dr; 
		
		return s;
	
	def doInit (self, obj) :
		#set in request
		obj.set("user", obj.get("ps.user"));
		
	#called as collection 
	def addItem (self, obj) :
		params = obj.getParams();
		recItem = params.get("item");
		collectionPath = params.getRef("collection");
		displayControl = params.get("displayControl.control");
		if recItem is not None  :       
			typePath = displayControl.getRef(".");
			type = itemtype.ItemType(typePath);
			#Pebble layout = type.get("layout", true);
			#prunedItem = self.appInst.pruneItem(recItem, layout);
	
			uniqueName = recItem.getValue("username");
			if uniqueName is not None :
				recItem.setTagName(uniqueName);
	
			pathAnalyzer = pathanalyzer.PathAnalyzer(collectionPath);
			parentForm = pathAnalyzer.getParentItemType();
			collectionField = parentForm.getField(collectionPath[collectionPath.rfind(".") + 1:]);
			typeArrayMaps = collectionField.get("type.typeArrayMaps");
			recType = collectionField.getRef("type.arrayFormId");
			form = itemtype.ItemType(recType);
	
			doc = docmodel.DocModel.getDatasource(collectionPath).create(collectionPath, recItem, typeArrayMaps, form);
	
			pebble.set("appIndexItem", doc.get("appIndexItem"));
	
			#notify
	
	def updateItem (self, obj) :
		params = obj.getParams();
		recItem = params.get("item");
		itemPath = params.getRef("itemPath");
		collection = itemPath[:itemPath.rfind(".")];
		name = itemPath[itemPath.lastIndexOf(".") + 1:];
		displayControl = params.get("displayControl.control");
		if recItem is not None :
			pathAnalyzer = pathanalyzer.PathAnalyzer(itemPath);
	
			typePath = displayControl.getRef(".");
			type = itemtype.ItemType(typePath);
			#Pebble layout = type.get("layout", true);
			#prunedItem = self.appInst.pruneItem(recItem, layout);
	
	#		ArrayList<Pebble> items = prunedItem.getRecords(".");
	#		for (Pebble item : items) {
	#		String meta;
	#		if path is not None :
	#		meta = path + "." + item.getMetaId();
	#		else:
	#		meta = item.getMetaId();
	#		}
	#		obj.set(meta, item);
	#		}
	
			relPath = pathAnalyzer.getRelPath();
			o = pathAnalyzer.getLastDoc();
			o.set(relPath, recItem);
	
			pa = pathanalyzer.PathAnalyzer(collection);
			typeArrayMaps = pa.getLastFieldType().get("typeArrayMaps");
			arrayForm = pa.getLastFieldType().get("arrayFormId");
	
			docmodel.DocModel.getDatasource(collection).update(collection, name, o, typeArrayMaps, itemtype.ItemType(arrayForm.getRef(".")));
	
	
			#TODO: pebble.set("appIndexItem", appIndexItem);
	
	def deleteItem (self, obj) :
		params = obj.getParams();
		appPath = params.getRef("appPath");
	
